### PR TITLE
domain,session: Store and delete internal session in the session pool

### LINKED ISSUE #34154: internal txn failed due to do gc
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

`tpcc prepare 1000 warehouse，run threads 1024，check gc safepoint and auto analyze start ts from tidb.log`

### 2. What did you expect to see? (Required)

`gc safepoint <= auto analyze start ts，and auto analyze internal txn success.`

### 3. What did you see instead (Required)

```
auto analyze internal txn failed due to do gc. 
tidb.log:
sh-4.4# tail -f /var/lib/tidb/log/tidb.log |grep "auto analyze"             
[2022/04/21 08:24:13.219 +00:00] [INFO] [analyze.go:146] ["analyze table `tpcc`.`customer` has failed"] [partition=] ["job info"="auto analyze columns"] ["start time"=2022/04/21 08:09:26.641 +00:00] ["end time"=2022/04/21 08:24:13.104 +00:00] [cost=14m46.463454126s]
[2022/04/21 08:24:13.369 +00:00] [INFO] [analyze.go:146] ["analyze table `tpcc`.`customer` has failed"] [partition=] ["job info"="auto analyze index PRIMARY"] ["start time"=2022/04/21 08:23:59.311 +00:00] ["end time"=2022/04/21 08:24:13.219 +00:00] [cost=13.908321366s]
[2022/04/21 08:24:13.425 +00:00] [INFO] [analyze.go:146] ["analyze table `tpcc`.`customer` has failed"] [partition=] ["job info"="auto analyze index idx_customer"] ["start time"=2022/04/21 08:24:10.107 +00:00] ["end time"=2022/04/21 08:24:13.369 +00:00] [cost=3.262278939s]
[2022/04/21 08:24:13.425 +00:00] [ERROR] [update.go:1182] ["[stats] auto analyze failed"] [sql="analyze table `tpcc`.`customer`"] [cost_time=14m48.909784259s] [error="[tikv:9006]GC life time is shorter than transaction duration, transaction starts at 2022-04-21 08:09:24.51 +0000 UTC, GC safe point is 2022-04-21 08:13:39.361 +0000 UTC"]
[2022/04/21 08:24:13.513 +00:00] [INFO] [update.go:1081] ["[stats] auto analyze triggered"] [sql="analyze table `tpcc`.`warehouse`"] [reason="too many modifications(254053/1000>0.5)"]
[2022/04/21 08:24:26.105 +00:00] [INFO] [analyze.go:146] ["analyze table `tpcc`.`warehouse` has finished"] [partition=] ["job info"="auto analyze columns"] ["start time"=2022/04/21 08:24:13.532 +00:00] ["end time"=2022/04/21 08:24:26.020 +00:00] [cost=12.488134647s]
[2022/04/21 08:24:27.132 +00:00] [INFO] [update.go:1081] ["[stats] auto analyze triggered"] [sql="analyze table `tpcc`.`stock`"] [reason="too many modifications(40981648/6.1637888e+07>0.5)"]


sh-4.4# cat /var/lib/tidb/log/tidb.log |grep "GetStartTSFromSession"
[2022/04/21 08:20:56.578 +00:00] [INFO] [session.go:3209] ["GetStartTSFromSession getting startTS of internal session"] [startTS=0] ["start time"=1970/01/01 00:00:00.000 +00:00]
[2022/04/21 08:23:56.534 +00:00] [INFO] [session.go:3209] ["GetStartTSFromSession getting startTS of internal session"] [startTS=0] ["start time"=1970/01/01 00:00:00.000 +00:00]
[2022/04/21 08:24:26.595 +00:00] [INFO] [session.go:3209] ["GetStartTSFromSession getting startTS of internal session"] [startTS=0] ["start time"=1970/01/01 00:00:00.000 +00:00]
[2022/04/21 08:40:56.586 +00:00] [INFO] [session.go:3209] ["GetStartTSFromSession getting startTS of internal session"] [startTS=0] ["start time"=1970/01/01 00:00:00.000 +00:00]
```

### 4. What is your TiDB version? (Required)

```
[2022/04/20 09:31:52.008 +00:00] [INFO] [printer.go:34] ["Welcome to TiDB."] ["Release Version"=v6.1.0-nightly] [Edition=Community] ["Git Commit Hash"=36c348899fc9d82db8760ad
[2022/04/20 09:31:52.009 +00:00] [INFO] [printer.go:48] ["loaded config"] [config="{\"host\":\"[0.0.0.0](http://0.0.0.0/)\",\"advertise-address\":\"[tc-tidb-1.tc](http://tc-tidb-1.tc/)-tidb-peer.testbed-endless-gc-gvj
[2022/04/20 09:31:52.009 +00:00] [INFO] [main.go:345] ["disable Prometheus push client"]                                                                                     
[2022/04/20 09:31:52.009 +00:00] [INFO] [store.go:74] ["new store"] [path=tikv://tc-pd:2379]
```

### PR DESCRIPTION
**Pull request by TonsnakeLin**
Signed-off-by: TonsnakeLin <lpbgytong@163.com>

### What problem does this PR solve?

Issue Number: close #34154

Problem Summary:

### What is changed and how it works?
We stored an internal session to the SessionManger and delete it in the function `(s *session) getInternalSession`, but not all internal sessions were got from it. We now store an internal session to the SessionManger and delete it from SessionPool's `get` and 'put' methods, so no internal sessions are lost.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by TonsnakeLin** [comment]
cc @cfzjywxk @sticnarf @youjiali1995 @MyonKeminta

**Comment by sticnarf** [comment]
Is it possible to cover it with a unit test?

**Comment by TonsnakeLin** [comment]
> Is it possible to cover it with a unit test?

I'll try it

**Comment by TonsnakeLin** [comment]
@cfzjywxk @sticnarf @youjiali1995 @MyonKeminta

File: session/session.go [hunk 1]
```diff
@@ -3206,6 +3200,11 @@ func GetStartTSFromSession(se interface{}) uint64 {
 	if txnInfo != nil {
 		startTS = txnInfo.StartTS
 	}
+
+	logutil.BgLogger().Info(
```
Comment (cfzjywxk) [inline session/session.go:3204]: This could be changed to `DEBUG` level.

File: domain/db_test.go [hunk 2]
```diff
@@ -56,3 +60,61 @@ func TestDomainSession(t *testing.T) {
 	_, err = se.Execute(context.Background(), "create global binding for select * from t where i>100 using select * from t use index(index_t) where i>100")
 	require.NoError(t, err)
 }
+
+func TestNormalSessionPool(t *testing.T) {
+	lease := 50 * time.Millisecond
```
Comment (cfzjywxk) [inline domain/db_test.go:65]: Increasing the lease maybe could prevent some weird failures.

**Review by sticnarf — COMMENTED** [review]
I have only one small question.

What if some error happens after `getInternalSession` gets a session from the pool? I find `getSnapshotInfoSchema` may raise an error.

It seems like a resource leak. But it seems to me that it does not affect calculating the min start TS, right?

**Comment by TonsnakeLin** [comment]
> 
Yes, I think you find a bug when use getInternalSession to get internal sessions from session pool. It does not affect calculating the min start TS, because the startTS of the session leaked is always 0.

> I have only one small question.
> 
> What if some error happens after `getInternalSession` gets a session from the pool? I find `getSnapshotInfoSchema` may raise an error.
> 
> It seems like a resource leak. But it seems to me that it does not affect calculating the min start TS, right?

**Review by sticnarf — APPROVED** [review]
I don't have more comments except those from @cfzjywxk

File: session/session.go (hunk 1 above)
Comment (TonsnakeLin) [reply session/session.go]: done

File: domain/db_test.go (hunk 2 above)
Comment (TonsnakeLin) [reply domain/db_test.go:65]: done

**Comment by sticnarf** [comment]
> Yes, I think you find a bug when use getInternalSession to get internal sessions from session pool. It does not affect calculating the min start TS, because the startTS of the session leaked is always 0.

Would you mind giving a quick fix here in this PR? It was not a real problem before, because the unused session should be GCed by the runtime, but here it is a real leak because it's saved to the map now.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**